PackageViewModel/SignPackageViewModel.cs

```diff
@@ -0,0 +1,363 @@
+                        collection,
+                        "Choose a Certificate for Package Signing",
+                        "Provide the code signing certificate for signing the package.",
+                        X509SelectionFlag.SingleSelection);
+
+                    if (certificates.Count > 0)
+                    {
+                        Certificate = certificates[0];
+                        CertificateFileName = null;
+                        ShowPassword = false;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                OnError(ex);
+            }
+        }
+
+        private void ShowCertificateCommandExecute()
+        {
+            var certificate = Certificate;
+            if (certificate != null)
+            {
+                try
+                {
+                    X509Certificate2UI.DisplayCertificate(certificate);
+                }
+                catch { }
+            }
+        }
+
+        // https://github.com/NuGet/NuGet.Client/blob/4c0c9658445573845ddbeff5656e4b3129f727a2/src/NuGet.Core/NuGet.Commands/SignCommand/SignCommandRunner.cs
+        public async Task<string> SignPackage()
+        {
+            ShowProgress = true;
+            CanSign = false;
+            Status = "Signing package...";
+
+            try
+            {
+                // change to AuthorSignPackageRequest when NuGet.Client is updated
+                using (var tempCertificate = new X509Certificate2(Certificate))
+                using (var signRequest = new SignPackageRequest(tempCertificate, GetHashAlgorithmName(tempCertificate)))
+                {
+                    SigningUtility.Verify(signRequest, NullLogger.Instance);
+
+                    var packagePath = _packageViewModel.GetCurrentPackageTempFile();
+                    var originalPackageCopyPath = Path.GetTempFileName();
+
+                    File.Copy(packagePath, originalPackageCopyPath, overwrite: true);
+
+                    using (var packageReadStream = File.OpenRead(packagePath))
+                    using (var packageWriteStream = File.Open(originalPackageCopyPath, FileMode.Open))
+                    using (var package = new SignedPackageArchive(packageReadStream, packageWriteStream))
+                    {
+                        var signer = new Signer(package, new X509SignatureProvider(null));
+                        await Task.Run(() => signer.SignAsync(signRequest, NullLogger.Instance, CancellationToken.None));
+                    }
+
+                    return originalPackageCopyPath;
+                }
+            }
+            catch (Exception e)
+            {
+                OnError(e);
+                return null;
+            }
+            finally
+            {
+                ShowProgress = false;
+                CanSign = true;
+            }
+        }
+
+        private void ValidateCertificate()
```

Comment:
I'm not completely following why this needs a timer with a semaphore?

Reply:
Currently this is run on every keypress in the password field.
And the validation might take some time which resulted in an invalid state.